cmd/backup/backup.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sort"
+	"time"
+
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/openshift/openshift-azure/pkg/log"
+	"github.com/openshift/openshift-azure/pkg/util/managedcluster"
+)
+
+const (
+	cMaxBackups = 6
+	copyRetries = 3
+	interval    = 5 * time.Second
+)
+
+var (
+	logLevel    = flag.String("loglevel", "Debug", "Valid values are Debug, Info, Warning, Error")
+	blobName    = flag.String("blobname", "", "Name of the blob")
+	group       = flag.String("group", "hourly", "The Blob will be called etcd/backup-<group>-<timestamp>")
+	destination = flag.String("destination", "", "Where to place the blob on the filesystem")
+	gitCommit   = "unknown"
+)
+
+type backupCmd struct {
+	abs     *azstorage.BlobStorageClient
+	ctx     context.Context
+	kubecli kubernetes.Interface
+}
+
+func (b *backupCmd) handleBackup(path string, endpoints []string, clientTLSSecret, namespace string) error {
+	tlsConfig, err := generateTLSConfig(b.kubecli, clientTLSSecret, namespace)
+	if err != nil {
+		return err
+	}
+
+	bm := NewBackupManagerFromWriter(b.kubecli, NewABSWriter(b.abs), tlsConfig, endpoints, namespace)
+
+	rev, etcdVersion, err := bm.SaveSnap(b.ctx, path)
+	if err != nil {
+		return fmt.Errorf("failed to save snapshot (%v)", err)
+	}
+	logrus.Infof("SaveSnap rev:%v etcdVersion:%v", rev, etcdVersion)
+	return nil
+}
+
+func (b *backupCmd) handleBackupPrune() error {
+	containerRef := b.abs.GetContainerReference("etcd")
+	containerExists, err := containerRef.Exists()
+	if err != nil {
+		return err
+	}
+	if !containerExists {
+		return fmt.Errorf("container %v does not exist", "etcd")
+	}
+
+	params := azstorage.ListBlobsParameters{
+		Prefix:     "backup-" + *group,
+		MaxResults: cMaxBackups + 5,
+	}
+	blobs, err := containerRef.ListBlobs(params)
+	if err != nil {
+		return err
+	}
+	toDelete := 0
+	if len(blobs.Blobs) > cMaxBackups {
+		toDelete = len(blobs.Blobs) - cMaxBackups
+	} else {
+		return nil
+	}
+	// this should sort oldest first
+	sort.SliceStable(blobs.Blobs, func(i, j int) bool {
+		return time.Time(blobs.Blobs[i].Properties.LastModified).Before(time.Time(blobs.Blobs[j].Properties.LastModified))
+	})
+	for _, blob := range blobs.Blobs {
```

Comment:
for robustness I would be highly inclined to filter blobs to only the ones you want to delete, then iterate them - the toDelete/continue thing is kinda nasty.